## Importing the dedpendencies

In [79]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler # we need to standardise the data to a common range
from sklearn.model_selection import train_test_split
from sklearn import svm # Support Vector Machine Algorithm
from sklearn.metrics import accuracy_score

Data Collection and Analysis
PIMA Diabetes Dataset (Kaggle)

In [80]:
# loading dataset to pandas dataframe
diabetes_dataset = pd.read_csv('/content/diabetes.csv')

In [81]:
diabetes_dataset.head()

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1

[5 rows x 9 columns]

Paramaters : Pregnancies	Glucose	BloodPressure	SkinThickness	Insulin	BMI	DiabetesPedigreeFunction	Age


Label = outcome


In [82]:
#printing rows and columns
diabetes_dataset.shape

(768, 9)

In [83]:
# getting the statistical measures of the dataframe
diabetes_dataset.describe()

Pregnancies     Glucose  ...         Age     Outcome
count   768.000000  768.000000  ...  768.000000  768.000000
mean      3.845052  120.894531  ...   33.240885    0.348958
std       3.369578   31.972618  ...   11.760232    0.476951
min       0.000000    0.000000  ...   21.000000    0.000000
25%       1.000000   99.000000  ...   24.000000    0.000000
50%       3.000000  117.000000  ...   29.000000    0.000000
75%       6.000000  140.250000  ...   41.000000    1.000000
max      17.000000  199.000000  ...   81.000000    1.000000

[8 rows x 9 columns]

In [84]:
diabetes_dataset['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

In [1]:
# Mean of all the values in label 0 and 1
diabetes_dataset.groupby('Outcome').mean()
# when we did this, we saw something
# we saw that after we found mean is 
# the person who has diabetes has more age than the person without diabetes
# and the glucose level of the person who has diabetes is higher than the person who doesn't has diabetes
# SO yeah, we didi this for our analysis

NameError: name 'diabetes_dataset' is not defined

In [86]:
# separating data and labels
x = diabetes_dataset.drop(columns='Outcome', axis=1)
y = diabetes_dataset['Outcome']


In [87]:
print(x)
y

     Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
0              6      148             72  ...  33.6                     0.627   50
1              1       85             66  ...  26.6                     0.351   31
2              8      183             64  ...  23.3                     0.672   32
3              1       89             66  ...  28.1                     0.167   21
4              0      137             40  ...  43.1                     2.288   33
..           ...      ...            ...  ...   ...                       ...  ...
763           10      101             76  ...  32.9                     0.171   63
764            2      122             70  ...  36.8                     0.340   27
765            5      121             72  ...  26.2                     0.245   30
766            1      126             60  ...  30.1                     0.349   47
767            1       93             70  ...  30.4                     0.315   23

[76

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

## Standardization of Data - Data Standardization

In [88]:
# BP can be around 60-70
# BMI can be around 26-36
# Glucose can be 90-130

# for standaridizing our data we have imported StandardScalar from sklearn.preprocessing


In [89]:
scalar = StandardScaler()

In [90]:
scalar.fit(x)

StandardScaler()

In [91]:
# Now transforming the data
standardized_data = scalar.transform(x)
# what it does is, it takes the data nd removes all the inconsistent values and it keeps all the values in a common range

In [92]:
# standardised_data = scalar.fit_transform(x)
# This fits the data and does the standardization in a single step

In [93]:
print(standardized_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [94]:
# Now all the data have been in the similar range

In [95]:
x = standardized_data
y = diabetes_dataset['Outcome']
# now we have have all the data in x and all the data in y
print(x)
print(y)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]
0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


In [96]:
x.shape

(768, 8)

In [97]:
# Now we need to split our data into training data and testing data


## Train Test Split


In [98]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, stratify = y, random_state=2)

SO here what have done is -
we have x = data above right, so what this train_test_split() will do is, it will split the data into two parts - \
training data - x_train \
testing data - x_test \
our model will be trained on the x_train and the once the model is train we will evaluate our model on the test data\
here the test_size = 0.2, it means 20% of the data is given to evalute the data\
y_train is the label for the x_train data

In [99]:
print(x.shape, x_train.shape, x_test.shape)

(768, 8) (614, 8) (154, 8)


Training the Model

In [100]:
# SVC = Support Vector Classifier
classifier = svm.SVC(kernel='linear')
#now this will load the svm model in the variable named classifier
# Now we will fit our traning data in this classifier

In [101]:
classifier.fit(x_train, y_train) # data and label

SVC(kernel='linear')

Evaluation of out model\
Evaluation is basically to check how many times our model is predicting it correctly.

In [102]:
# Model Evaluation
# so first finding the accuracy score on the training data
# and we will predict the labels for it and then we will compare our model prediction and our actual values
# this is how we are going to get the accuracy score muehehehehe
# accuracy score on the training data

x_train_prediction = classifier.predict(x_train)
training_accuracy = accuracy_score(x_train_prediction, y_train)
print("Accuracy score of training data : ",training_accuracy)

Accuracy score of training data :  0.7866449511400652


In [103]:
# accuracy score of the test data
x_test_prediction = classifier.predict(x_test)
testing_accuracy = accuracy_score(x_test_prediction, y_test)
print("Accuracy of the testing data : ", testing_accuracy)

Accuracy of the testing data :  0.7727272727272727


In [104]:
# Making a predictive system
input_data = (6,148,72,35,0,33.6,0.627,50)
# changing the input data to numpy array
input_data_as_numpy_array = np.asarray(input_data)
# reshape the data as we are predicting are one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)

## Why did we reshape?
Our model is trained on 768 examples and there are 8 columns\
and as an input we are only giving one data point, and our model expects the input of all 768 values\
This reshape() tell the model that we are only giving one datapoint as input and we only need output for only one datapoint.

In [105]:
# there is one more problem, we cannot give the values like our input data to the input_data.
# because while training the data we standardised the data and
# hence we have to standardise the input data aswell otherwise our model won't be able to make the prediction properly.

In [106]:
input_data_standardized = scalar.transform(input_data_reshaped)
print(input_data_standardized)

[[ 0.63994726  0.84832379  0.14964075  0.90726993 -0.69289057  0.20401277
   0.46849198  1.4259954 ]]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [108]:
prediction = classifier.predict(input_data_standardized)
print(prediction)

[1]


In [111]:
if(prediction[0]==1): # we did [0] because prediction is a list
  print("Person is Diabetic")
else:
  print("Person is not Diabetic")

Person is Diabetic
